pyccel/stdlib/cuda_ndarrays/cuda_ndarrays.cu

```diff
@@ -0,0 +1,86 @@
+#include "cuda_ndarrays.h"
+
+void    device_memory(void** devPtr, size_t size)
+{
+    cudaMalloc(devPtr, size);
+}
+
+void    host_memory(void** devPtr, size_t size)
+{
+    cudaMallocHost(devPtr, size);
+}
+t_ndarray   cuda_array_create(int32_t nd, int64_t *shape, enum e_types type, bool is_view ,
+enum e_memory_locations location)
+{
+    t_ndarray  arr;
+    void (*fun_ptr_arr[])(void**, size_t) = {host_memory, device_memory};
+
+    arr.nd = nd;
+    arr.type = type;
+    switch (type)
+    {
+        case nd_int8:
+            arr.type_size = sizeof(int8_t);
+            break;
+        case nd_int16:
+            arr.type_size = sizeof(int16_t);
+            break;
+        case nd_int32:
+            arr.type_size = sizeof(int32_t);
+            break;
+        case nd_int64:
+            arr.type_size = sizeof(int64_t);
+            break;
+        case nd_float:
+            arr.type_size = sizeof(float);
+            break;
+        case nd_double:
+            arr.type_size = sizeof(double);
+            break;
+        case nd_bool:
+            arr.type_size = sizeof(bool);
+            break;
+    }
+    arr.is_view = is_view;
+    arr.length = 1;
+    cudaMallocManaged(&(arr.shape), arr.nd * sizeof(int64_t));
+    for (int32_t i = 0; i < arr.nd; i++)
+    {
+        arr.length *= shape[i];
+        arr.shape[i] = shape[i];
+    }
+    arr.buffer_size = arr.length * arr.type_size;
+    cudaMallocManaged(&(arr.strides), nd * sizeof(int64_t));
+    for (int32_t i = 0; i < arr.nd; i++)
+    {
+        arr.strides[i] = 1;
+        for (int32_t j = i + 1; j < arr.nd; j++)
+            arr.strides[i] *= arr.shape[j];
+    }
+    if (!is_view)
+        (*fun_ptr_arr[location])(&(arr.raw_data), arr.buffer_size);
+    return (arr);
+}
+
+int32_t cuda_free_host(t_ndarray  arr)
+{
+    if (arr.shape == NULL)
+        return (0);
+    cudaFreeHost(arr.raw_data);
```

Comment:
I think it would be better to call free directly here
https://forums.developer.nvidia.com/t/cudafreehost-consistently-20x-slower-than-free-cudafree-full-runnable-example-code-available/217706

Reply:
Yes, you are totally right. I ran a simple test to compare the performance of malloc/free and cudaMallocHost/cudaFreeHost. I found that cudaMallocHost/cudaFreeHost is approximately 3000 times slower than malloc/free.
```c
int main(void) {
    // Test malloc/free
    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            char *a = (char*)malloc(1);
            free(a);
        }
    }

    // Test cudaMallocHost/cudaFreeHost
    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            char *a;
            cudaMallocHost(&a, 1);
            cudaFreeHost(a);
        }
    }
}
```

Reply:
I'm surprised that it is that bad! I wonder if the compiler is capable of optimising away the malloc/free call but not the cudaMallocHost/cudaFreeHost call, you might want to test using the array

Reply:
Please What do you mean Emily ??

Reply:
[0e9292e](https://github.com/pyccel/pyccel-cuda/pull/67/commits/0e9292e82b815d95c17d20533e8b5b98f27203a5)